File: src/modules/graphs/3_cycleDetection_Undirected/DFS.ts
export class Solution {
    checkCycle(node: number, parent: number, adj: number[][], vis: number[]): boolean {
        vis[node] = 1;
        for (const neighbor of adj[node]) {
            if (!vis[neighbor]) {
                if (this.checkCycle(neighbor, node, adj, vis)) return true;
            } else if (neighbor !== parent) return true;
        }
        return false;
    }

    isCycle(Adj: number[][]) {
        const n = Adj.length;
        const visited = new Array(n).fill(0);
        for (let i = 0; i < n; i++) {
            if (!visited[i]) {
                if (this.checkCycle(i, -1, Adj, visited)) return true;
            }
        }
        return false;
    }
}

File: src/modules/graphs/3_cycleDetection_Undirected/dfs.go
package main

func isCycle(Adj [][]int) bool {
	n := len(Adj)
	visited := make([]bool, n)

	for i := 0; i < n; i++ {
		if !visited[i] {
			if checkCycleDFSRecursive(i, -1, Adj, visited) {
				return true
			}
		}
	}

	return false
}

func checkCycleDFSRecursive(start int, parent int, Adj [][]int, visited []bool) bool {
	visited[start] = true
	for _, neighbor := range Adj[start] {
		if !visited[neighbor] {
			if checkCycleDFSRecursive(neighbor, start, Adj, visited) {
				return true
			}
		} else if neighbor != parent {
			return true
		}
	}
	return false
}
